dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/ErrorDecoder.java

```diff
@@ -117,4 +133,54 @@ private static String toString(InputStream body) throws IOException {
             return CharStreams.toString(reader);
         }
     }
+
+    private static ResponseDiagnostic diagnostic(Response response) {
+        return new ResponseDiagnostic(diagnosticArgs(response));
+    }
+
+    private static ImmutableList<Arg<?>> diagnosticArgs(Response response) {
+        ImmutableList.Builder<Arg<?>> args = ImmutableList.<Arg<?>>builder().add(SafeArg.of("status", response.code()));
+        recordHeader(HttpHeaders.SERVER, response, args);
+        recordHeader(HttpHeaders.CONTENT_TYPE, response, args);
+        recordHeader(HttpHeaders.CONTENT_LENGTH, response, args);
+        recordHeader(HttpHeaders.CONNECTION, response, args);
+        recordHeader(HttpHeaders.DATE, response, args);
+        recordHeader("x-envoy-response-flags", response, args);
+        recordHeader("x-envoy-response-code-details", response, args);
+        return args.build();
+    }
+
+    private static void recordHeader(String header, Response response, ImmutableList.Builder<Arg<?>> args) {
+        response.getFirstHeader(header).ifPresent(server -> args.add(SafeArg.of(header, server)));
+    }
+
+    private static final class ResponseDiagnostic extends RuntimeException implements SafeLoggable {
```

Comment:
Looks like second time we're using this trick, shall we extract a little util thingie

Reply:
They're used in separate modules, I don't think we want to expose public API for something this simple.